### ISSUE TITLE
Question: Custom optimizer for NeuralNetClassifier

### ISSUE THREAD

**Issue by supetronix** [skorch-dev/skorch#696]
Hi,

I looked through documentation, code samples and StackOverflow but I didn't find any easy/quick solution (besides subclassing NuralNet) to use a custom optimizer. 

**Case:**
Instead of using a PyTorch optimizer class name, e.g.

`optimizer=torch.optim.Adam,`

I want to use a custom "Lookahead" optimizer that will take "Adam" as input.

Code sample for PyTorch:
```
        adam_optimizer = torch.optim.Adam(self.parameters(), lr = self.cfg.lr, 
                                     weight_decay=self.cfg.weight_decay)
        optimizer = Lookahead(optimizer=adam_optimizer,k=5,alpha=0.5)
```

code for lookahead optimizer is from:
[Lookahead github](https://github.com/lonePatient/lookahead_pytorch)

any help with easy solution will be appreciated. Or it could be added as a future enhancement to allow passing custom optimizer.

**Comment by BenjaminBossan** [comment]
The problem here is not that you want to pass a custom optimizer, it is that your custom optimizer has a different signature than the other PyTorch optimizers. The solution is hence to take control of how the arguments are routed. Could you try if this works:

```python
def make_lookahead(parameters, optimizer_cls, k, alpha, **kwargs):
    optimizer = optimizer_cls(parameters, **kwargs)
    return Lookahead(optimizer=optimizer, k=k, alpha=alpha)

net = NeuralNet(
    ...,
    optimizer=make_lookahead,
    optimizer__optimizer_cls=torch.optim.Adam,
    optimizer__weight_decay=cfg.weight_decay,
    lr=cfg.lr,
)
```

The idea is basically to have a function that takes care of passing some parameters to `Lookahead` and the others to the optimizer class (i.e. `Adam`).

The alternatives would be to 1) subclass `NeuralNet` and override `initialize_optimizer` (as you mentioned) or 2) refactor `Lookahead` so that instead of requiring the initialized optimizer, it accepts the optimizer class and initializes it inside of its `__init__`.

**Comment by supetronix** [comment]
Thanks @BenjaminBossan , your suggested solution worked perfectly (i.e. create a method to encapsulate Lookahead optimizer). 

Is it documented somewhere on Skorch website? I tried to look through documentation before asking this question so a link may be helpful for me and anyone else that may come across this questions.

Not to pile up multiple questions on same issue but I couldn't find any "learning rate finder" method in Skorch documentation, curious if you have any link on that too. 

Example is FastaiLRFinder:
https://pytorch.org/ignite/_modules/ignite/contrib/handlers/lr_finder.html

**Comment by BenjaminBossan** [comment]
> Is it documented somewhere on Skorch website? I tried to look through documentation before asking this question so a link may be helpful for me and anyone else that may come across this questions.

I don't think it's documented, I'm also not quite sure where and how to document it, since it's rather specific. If you have a suggestion, feel free to post it or even create a PR.

> I couldn't find any "learning rate finder" method in Skorch documentation, curious if you have any link on that too.

No, we don't. I looked at the ignite implementation but it's very ignite specific and wouldn't work with skorch. Basically, we'd need our own implementation.

**Comment by supetronix** [comment]
how about modifying the "Optimizer" section on:
https://github.com/skorch-dev/skorch/blob/92ae54ba7cb636488cf42cc85894d52765f6527a/docs/user/neuralnet.rst

I am thinking if we can change first few words from:
This should be a PyTorch optimizer

to
This should be a PyTorch optimizer or a python function that will return a PyTorch optimizer

and then give example of this function below it.

I can create a pull request if this makes sense.

**Comment by BenjaminBossan** [comment]
Okay, thanks for taking this.

Instead of modifying the first few words of that section, I would propose to instead include your example below the existing example. The reason for this is that otherwise, I would be afraid that the first paragraph will contain too much information.
